docs/guides/spmd_api.md

```diff
@@ -0,0 +1,429 @@
+<!-- #endregion -->
+
+```python id="_noFQzrgJ2j0"
+class MLP(nn.Module):
+  num_layers: int
+  depth: int
+  use_scan: bool
+  @nn.compact
+  def __call__(self, x):
+    if self.use_scan:
+      x, _ = nn.scan(DotReluDot, length=self.num_layers, 
+                    variable_axes={"params": 0},
+                    split_rngs={"params": True},
+                    metadata_params={nn.PARTITION_NAME: None}
+                    )(self.depth)(x)
+    else:
+      for i in range(self.num_layers):
+        x, _ = DotReluDot(self.depth)(x)
+    return x
+```
+
+<!-- #region id="pZ6E5btPr5yZ" -->
+# Sharding specification
+
+Now we want to generate the `PartitionSpec` that `pjit` should receive as annotations of input and output data.
+
+The input is relatively simple. For data parallelism, we can shard the batched input `x`s across the `data` axis, by denoting the batch axis as `data`.
+
+<!-- #endregion -->
+
+```python id="63TwLe6vNtID"
+# Input partition rule - data-parallel only
+x_spec = PartitionSpec('data', None)  # dimensions: (batch, length)
+x_spec
+```
+
+<!-- #region id="3d6b15d08980" -->
+However, to generate `PartitionSpec` for the output, we need to use some actual output as reference. One solution is to create a model and evaluate `model.init` abstractly using `jax.eval_shape`, and then use `nn.get_partition_spec` to automatically generate the `PartitionSpec`. 
+
+The code below shows how to get the output spec if you use `flax.training.train_state` to carry out your initialization and training steps, in which case your `pjit`ted function will output a `TrainState`.
+
+> A side note: Here we define our `init_fn` as purely functional and takes `model` and `optimizer` as arguments. This is not necessary - you can simply define with `def init_fn(k, x):` and all will work fine here.
+> 
+> This guide doesn't do it because later we will show you another way to define your model and will run `init_fn` with another model instance. But this does a bit of trouble because `jax.eval_shape` only takes numeric inputs, and we have to create an abstract closure before feeding the function in.
+<!-- #endregion -->
+
+```python id="YqSxbGkxo4QV"
+# MLP hyperparams
+B, L, E, S = 8, 4, 1024, True
+# fake inputs
+x = jnp.ones((B, E))
+k = random.PRNGKey(0)
+
+optimizer = optax.adam(learning_rate=0.001)
+model = MLP(L, E, S)
+
+def init_fn(k, x, model, optimizer):
+  variables = model.init(k, x)
+  state = train_state.TrainState.create(
+    apply_fn=model.apply,
+    params=variables['params'],
+    tx=optimizer)
+  return state
+
+# make an abstract closure here because jax.eval_shape only takes numeric values
+init_fn_abstract = functools.partial(init_fn, model=model, optimizer=optimizer)
+with mesh:
+  abstract_variables = jax.eval_shape(init_fn_abstract, k, x)
+# we need a rule on every leaf of the input variables, in this case a nested dict.
+state_spec = nn.get_partition_spec(abstract_variables)
+state_spec
+```
```

Comment:
The output shows now three dimensions for the `PartitionSpec`, while you only provide 2 as input. I suppose this is because of the scan, but it is a but surprising so it might be worth clarifying this.

Reply:
The PartitionSpec shows that the params W1 and W2 has three dimensions, which WAI. The output of both pjitted functions are `TrainState`, which contains the params.